Merge TDA/fxa-geodb with Mozilla/fxa-geodb

TDA opened the pull request:
@vladikoff 
Found a way to keep commits woohoo!! 💃

vladikoff commented on README.md:
```diff
@@ -1,2 +1,63 @@
-# fxa-geodb
```
We also need a `LICENSE` file like https://github.com/mozilla/fxa-content-server/blob/master/LICENSE

TDA replied:
Added.

shane-tomlinson replied:
Can you add a LICENSE section to this doc that just points to the license file? See https://github.com/mozilla/fxa-content-server/#license

vladikoff commented on README.md:
````diff
@@ -1,2 +1,63 @@
+Clone/fork the repo and run `npm i`. Then, include the module in your source file, like so:
+
+```JavaScript
+var GeoDB = require('./fxa-geodb');
````
> Clone/fork the repo and run `npm i`. Then, include the module in your source file, like so:

this usage seems unclear. how is this module consumed, via npm?

vladikoff replied:
should `GeoDB` start with a capital letter?

TDA replied:
According to @shane-tomlinson , Its going to be a node module. IIRC this was so that we can reuse this for other projects, such as customs-server.

vladikoff replied:
Should I be allowed to set `sources` and other config options, such as 
```
var geoDb = require('fxa-geodb')({
  sources: require('../sources.json'),
  ....
})

vladikoff commented on README.md:
```diff
@@ -1,2 +1,63 @@
+GeoDB(ip).then(function (location) {
+		// success, resolved
+		// location data is available here
+      }, function (err) {
```
really strange indent here.

vladikoff commented on README.md:
````diff
@@ -1,2 +1,63 @@
+      	// Uh-oh error
+      });
+```
+where `timestamp` can be any JavaScript compatible representation of time.
````
What does the `timestamp` do? 

> compatible representation of time.

`Date()` object?

TDA replied:
Yup!!

vladikoff replied:
should probably just say that / give example of timestamp. Also not clear what `timestamp` does

TDA replied:
`timestamp` is given so that we can send an email with custom timestamp, say, after a week or so. Was suggested by @rfk 
Sure!, I will update with an example of a timestamp.

vladikoff replied:
`timestamp` does nothing for maxmind integration here it seems. IMHO it should be left out.

TDA replied:
From IRC:
<saipc> reason being time needs to be localized according to ip
<saipc> we need the location data to get the right time, so if ip was in NY, USA, it would need to be done as timestamp(utc-5)

rfk replied:
Does the information returned here allow some other component to localize a timestamp for display purposes?  If so then we probably don't need to pass in the timestamp, and can just return e.g. the timezone to use.

(From memory, we talked about having this library generate the actual localized string for use in the email, in which case it definitely would need to know the timestamp to include.  But that's less important if it's just returning raw data that will be rendered somewhere else).

vladikoff commented on tasks/require.js:
```diff
@@ -0,0 +1,30 @@
+module.exports = function (grunt) {
+  'use strict';
+
+  grunt.config('requirejs', {
```
why do we need `requirejs`?

vladikoff commented on scripts/max-loader.js:
```diff
@@ -0,0 +1,73 @@
+log('Downloading to %s', target);
+
+// import the list of files to download
+var sources = require('../sources.json');
```
it should allow me to configure custom sources

TDA replied:
Do you mean it should let you configure the name/path of the sources.json file, or configure the files that need to be downloaded? The latter is already available. 
is there a use case for the former?

vladikoff replied:
configure the whole thing, I should be able to pass the whole sources.json that you below.

TDA replied:
Will add configuration support.

vladikoff commented on scripts/max-loader.js:
```diff
@@ -0,0 +1,73 @@
+startDownload(work);
+
+// after that run periodically
+new CronJob('30 30 1 * * 3', function() {
```
instead of `magical` CronJob update,  the module should provide an `.update()` (or better named) call, that either takes options or that uses options provided in the constructor. then we can:

1. call `.update()` on server start to download the zip if geolocation feature is enabled and file is not present.
2. create a `.js` (update-locale.js) script that calls `.update()` in it and updates sources.

TDA replied:
After discussion with Shane and Vlad: we need to provide a function that is configurable by the end-user.

shane-tomlinson replied:
I was going to write something nearly identical to what @vladikoff already did. Adding tests for the extracted .js file would also be a big bonus.

vladikoff commented on scripts/max-loader.js:
```diff
@@ -0,0 +1,73 @@
+}
+
+function log() {
+  console.log.apply(console, arguments);
```
we probably need `mozlog` support in some capacity (https://github.com/mozilla/mozlog) if possible, such that the `last update` messages are logged properly to Heka

vladikoff commented on README.md:
```diff
@@ -1,2 +1,63 @@
+--
+### Updating
+A Cron job that runs every week on Wednesday at 1:30:30 AM updates the Geodata-DB from Maxmind. 
+All you have to do while using the repo for the first time is to run `npm run-script build`, and the 
```
instead updating should use `2.` from https://github.com/mozilla/fxa-geodb/pull/1/files#r68667768

vladikoff commented on sources.json:
```diff
@@ -0,0 +1,3 @@
+{
+  "cities-db.mmdb.gz": "http://geolite.maxmind.com/download/geoip/database/GeoLite2-City.mmdb.gz"
```
can this use HTTPS?

TDA replied:
Absolutely, I will change it.

vladikoff commented on src/fxa-geodb.js:
```diff
@@ -0,0 +1,79 @@
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+const db = '../db/cities-db.mmdb';
+const db_backup = '../db/cities-db.mmdb-backup';
```
all these probably need to be configurable paths, @jbuck for docker purposes, correct me if wrong.

TDA replied:
Absolutely, will make this configurable.

vladikoff commented on src/fxa-geodb.js:
```diff
@@ -0,0 +1,79 @@
+const Promise = require('bluebird');
+
+function GeoDB(ip, time_stamp) {
+  'use strict';
```
this code below should probably just use `return new Promise(function(resolve, reject) {`instead of calling `return Promise.reject({` everywhere

TDA replied:
Fixed now :)

vladikoff commented on src/fxa-geodb.js:
```diff
@@ -0,0 +1,79 @@
+
+  // at this point, we know ip is a string and is
+  // non-empty, we can validate it
+  if (! maxmind.validate(ip)) {
```
from above `the ip is defined, is a string, and is not empty` does `maxmind.validate(` not do that for us already?

TDA replied:
Yes, it does! Removed.

vladikoff commented on src/fxa-geodb.js:
```diff
@@ -0,0 +1,79 @@
+    },
+    time_zone: cityData.location.time_zone,
+    // can remove this if not required
+    cityData: cityData
```
there is a mix of casing in this object, do we want that?

TDA replied:
This won't be returned Vlad, I added it so that you guys can know what is the *entire* data that we get back, have removed it now.

vladikoff commented on src/fxa-geodb.js:
```diff
@@ -0,0 +1,79 @@
+    city: cityData.city.names.en,
+    continent: cityData.continent.names.en,
+    country: cityData.country.names.en,
+    local_time: new Date(time_stamp).toLocaleString('en', {timeZone: cityData.location.time_zone}),
```
should this always be `'en'`?

TDA replied:
After discussion with rfk and vlad, we are nixing this from the repo.

vladikoff commented on src/fxa-geodb-example.js:
```diff
@@ -0,0 +1,15 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
```
example should probably not be in `src`

vladikoff commented:
@TDA good start! A few small tweaks and it should be good to go.

 I'm mostly confused by the presence of requirejs in here, I don't think we need that task.

shane-tomlinson commented on examples/fxa-geodb-example.js:
```diff
@@ -0,0 +1,15 @@
+  const geodb = require('../src/fxa-geodb');
+  // New York timezone IP
+  geodb('128.192.8.8')
+    .then(function(city) {
```
is this supposed to be `city` or `location`?

TDA replied:
`location` it is. Sorry for the confusion.

vbudhram commented on test/fxa-geodb.js:
```diff
@@ -0,0 +1,75 @@
+      });
+  });
+
+  it('returns an object with location details when supplied with a valid ip address', function () {
```
Possible to have test cases where the IP address does not return the city/country information, since we still build out emails when this data is not present?

TDA replied:
Absolutely, I need to rewrite these tests to reflect that we are removing all validation (as Vlad suggested), and only using maxmind.validate :)

vbudhram commented on test/fxa-geodb.js:
```diff
@@ -0,0 +1,75 @@
+    };
+    return GeoDB(ip)
+      .then(function (location) {
+        assert.equal(location.country, 'United States', 'Country not returned correctly');
```
nit: Should the asserts below contain `not`? They are checking equal.

TDA replied:
@vbudhram Isn't that the failure message? So if it fails, it will print `Country not returned correctly` ?

vladikoff commented on scripts/max-loader.js:
```diff
@@ -0,0 +1,73 @@
+const requestProgress = require('request-progress');
+const mkdirp = require('mkdirp');
+const cp = require('child_process');
+const CronJob = require('cron').CronJob;
```
we might have to remove new stuff such as `const` because we might still be on node 0.10 when this is ready.

TDA replied:
Noted, will change back to `var`.

shane-tomlinson commented on scripts/max-loader.js:
```diff
@@ -0,0 +1,75 @@
+      cp.execFile('cp', [existing_file, existing_file_backup]);
+      // force overwrite, even if file exists already
+      // since we took a backup, this wont cause an issue (hopefully)
+      cp.execFile('gunzip', [target, '-f']);
```
what happens if `gunzip` is not available on the system?

shane-tomlinson commented on scripts/max-loader.js:
```diff
@@ -0,0 +1,75 @@
+});
+var log = mozlog();
+
+var target = path.join(process.cwd(), '../db');
```
should be:

```js
path.join(process.cwd(), '..', 'db');
```
path.join takes care of inserting the appropriate directory separator for the environment.

shane-tomlinson commented on scripts/max-loader.js:
```diff
@@ -0,0 +1,75 @@
+var requestProgress = require('request-progress');
+var mkdirp = require('mkdirp');
+var cp = require('child_process');
+var CronJob = require('cron').CronJob;
```
sort this list please

shane-tomlinson commented on scripts/max-loader.js:
```diff
@@ -0,0 +1,75 @@
+var CronJob = require('cron').CronJob;
+
+// set up mozlog, default is `heka`
+var mozlog = require('mozlog');
```
I would not run *any* of this code until some exported initialization function is called. Starting processing just by virtue of requiring the file is icky.

shane-tomlinson commented on scripts/max-loader.js:
```diff
@@ -0,0 +1,75 @@
+log.info('Downloading to %s', target);
+
+// import the list of files to download
+var sources = require('../sources.json');
```
```
require(path.join(__dirname, '..', 'sources.json'));
```

shane-tomlinson commented on scripts/max-loader.js:
```diff
@@ -0,0 +1,75 @@
+for (var source in sources) {
+  var url = sources[source];
+  log.info('adding ' + url);
+  work.push(download(url, path.join(target, source)));
```
what does the output target path look like?

shane-tomlinson commented on scripts/max-loader.js:
```diff
@@ -0,0 +1,75 @@
+      // target-backup (without .gz extension)
+      var existing_file = target.slice(0, -3);
+      var existing_file_backup = existing_file + '-backup';
+      cp.execFile('cp', [existing_file, existing_file_backup]);
```
`cp` assumes a unix based system, there must be some environment agnostic variant that can be used.

shane-tomlinson commented on scripts/max-loader.js:
```diff
@@ -0,0 +1,75 @@
+  };
+}
+
+function startDownload(work) {
```
what is work?

shane-tomlinson commented on scripts/max-loader.js:
```diff
@@ -0,0 +1,75 @@
+
+// import the list of files to download
+var sources = require('../sources.json');
+var work = [];
```
Can you give this a more descriptive name, something like `remainingDownloads`?

shane-tomlinson commented on scripts/max-loader.js:
```diff
@@ -0,0 +1,75 @@
+new CronJob('30 30 1 * * 3', function() {
+  // Cron job that runs every week on Wednesday at 1:30:30 AM.
+  startDownload(work);
+}, null, true, 'America/Los_Angeles');
```
Why `America/Los_Angeles`?

shane-tomlinson commented on src/fxa-geodb.js:
```diff
@@ -0,0 +1,85 @@
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+// these are defaults, can be overloaded by configuring options
+var db = __dirname + '/../db/cities-db.mmdb';
```
Use path.join to create the db output path. Give `db` a more descriptive name, otherwise readers will think it's a db object they can call functions on. This is really a `dbSourcePath` or similar.

shane-tomlinson replied:
`db` and `db_backup` are really `DEFAULT_DB_PATH` and `DEFAULT_BACKUP_DB_PATH` now that I see how they are used below.

shane-tomlinson commented on src/fxa-geodb.js:
```diff
@@ -0,0 +1,85 @@
+
+// these are defaults, can be overloaded by configuring options
+var db = __dirname + '/../db/cities-db.mmdb';
+var db_backup = __dirname + '/../db/cities-db.mmdb-backup';
```
Use camelCase to stay consistent. Again, give this a more descriptive name, this is really a path of some sort.

shane-tomlinson commented on src/fxa-geodb.js:
```diff
@@ -0,0 +1,85 @@
+        });
+      }
+
+      var city_lookup, city_data;
```
camelCase these too.

shane-tomlinson commented on src/fxa-geodb.js:
```diff
@@ -0,0 +1,85 @@
+  return function (ip) {
+    return new Promise(function (resolve, reject) {
+      // check if ip is valid
+      if (!maxmind.validate(ip)) {
```
I generally prefer spaces after `!`

shane-tomlinson commented on src/fxa-geodb.js:
```diff
@@ -0,0 +1,85 @@
+      // `db_backup` as free version or when `db` fails
+      // to load for some reason
+      try {
+        city_lookup = maxmind.open(db);
```
`city_lookup` seems to be misnamed, this is the actual db with functions that can be invoked.

shane-tomlinson commented on src/fxa-geodb.js:
```diff
@@ -0,0 +1,85 @@
+      // to load for some reason
+      try {
+        city_lookup = maxmind.open(db);
+        city_data = city_lookup.get(ip);
```
`city_data` also seems to be misnamed, `city` is very precise. This is more like `location` or `locationData`

shane-tomlinson commented on src/fxa-geodb.js:
```diff
@@ -0,0 +1,85 @@
+        }
+      }
+
+      if (city_data == null) {
```
`===` > `==`

shane-tomlinson commented on src/fxa-geodb.js:
```diff
@@ -0,0 +1,85 @@
+        });
+      }
+
+      // return an object with city, country, continent,
```
You are resolving with a function it looks like, not an object.

shane-tomlinson replied:
Why a function btw?

TDA replied:
`location` is indeed an object here. The function is a constructor that constructs the `location` object.
`typeof location === 'object';` // true

shane-tomlinson replied:
oh, I missed the `new`.  It seems like there's an easier way to do this:

```js
var location = {};
if (city_data.location) {
  location.accuracy = city_data.location.accuracy_radius;
  location.ll = {}
}
...
```
At a minimum, the intent will be clearer, and it really just requires changing `this` to `location` and replacing `new function () {` with `{`

TDA replied:
Sounds good, but I was just wanting to make sure that the object is clearly encapsulated. I am willing to change this if its causing confusion.

shane-tomlinson replied:
Another option is to move the entire constructor out of the middle of this function, and pass to the constructor the `city_data`:

```js
var location = new Location(city_data);
```

where Location is the function you currently have, moved to the outer level scope:

```js
function Location (city_data) {
  if (city_data.location) {
    this.accuracy = ...
  }
}
```

Don't forget about camelCase too. ;)

shane-tomlinson commented on src/fxa-geodb.js:
```diff
@@ -0,0 +1,85 @@
+      var location = new function () {
+        if (city_data.location) {
+          this.accuracy = city_data.location.accuracy_radius;
+          this.ll = {
```
`ll` leaves a bit to be desired, can this be `coordinates` or at a min `coords`?

shane-tomlinson commented on src/fxa-geodb.js:
```diff
@@ -0,0 +1,85 @@
+      }
+
+      if (city_data == null) {
+        reject({
```
It seems like once you reject, you should immediately return to abort any additional processing (in all cases)

shane-tomlinson commented on package.json:
```diff
@@ -0,0 +1,45 @@
+    "geodb",
+    "location"
+  ],
+  "author": "Sai Pc",
```
We have been using an `author` of `Mozilla (https://mozilla.org/)`.

shane-tomlinson commented on examples/fxa-geodb-example.js:
```diff
@@ -0,0 +1,20 @@
+
+(function() {
+  'use strict';
+  var geoDb = require('../src/fxa-geodb')({
```
Use path.join to create your paths here too.

jrgm replied:
Not necessary. Node `require()` has special handling - https://nodejs.org/dist/v0.10.46/docs/api/modules.html#modules_all_together

shane-tomlinson commented on README.md:
```diff
@@ -1,2 +1,49 @@
+
+--
+### Updating
+A Cron job that runs every week on Wednesday at 1:30:30 AM updates the Geodata-DB from Maxmind.
```
1:30:30 AM in which timezone? You can just prefix a `0` and skip the `AM`. I usually find it easier to write times in UTC. Folks worldwide can convert to their local timezone easily from that.

shane-tomlinson commented on README.md:
````diff
@@ -1,2 +1,49 @@
+Clone/fork the repo and run `npm i`. Then, include the module in your source file, like so:
+
+```JavaScript
+var geoDb = require('./fxa-geodb');
````
When this is an npm module, it'll just be:

```js
var geoDb = require('fxa-geodb');
```

shane-tomlinson commented on README.md:
```diff
@@ -1,2 +1,49 @@
+--
+### Updating
+A Cron job that runs every week on Wednesday at 1:30:30 AM updates the Geodata-DB from Maxmind.
+All you have to do while using the repo for the first time is to run `npm run-script build`, and the cron job will keep running in the background.
```
This could also use a "Getting involved", and "Submitting bugs" sections.

shane-tomlinson replied:
People who download the repo via npm are not going to read this doc about running `npm run-script build` to install the database. I would run the db-fetch tool as a postinstall script so consumers have to do no additional work.

shane-tomlinson commented on src/fxa-geodb.js:
```diff
@@ -0,0 +1,85 @@
+module.exports = function (options) {
+  'use strict';
+  options = options || {};
+  db = options.db || db;
```
Use locally defined variables here instead of overwriting those from the outer scope.

shane-tomlinson commented on src/fxa-geodb.js:
```diff
@@ -0,0 +1,85 @@
+      if (!maxmind.validate(ip)) {
+        reject({
+          message: ERRORS.IS_INVALID
+        });
```
return after all `rejects` to abort any further processing.

shane-tomlinson commented on src/fxa-geodb.js:
```diff
@@ -0,0 +1,85 @@
+        // if it failed with primary database,
+        // try with backup database
+        try {
+          city_lookup = maxmind.open(db_backup);
```
I see a couple of problems here. 

Keep in mind how and where this library is going to be used - the lookup function has the potential to be called many thousands of times an hour. Opening and reading the DB file synchronously for every query is going to be a performance bottleneck. Instead, open the DB on startup, keep it open for all queries. If the main DB file isn't available/valid, open the backup. If neither is usable, throw an error.

Next, I'm not sure there is any real value in querying the backup database if the info is not found in the primary database. I'd just stick with one for now.

shane-tomlinson commented on src/fxa-geodb.js:
```diff
@@ -0,0 +1,85 @@
+        }
+
+        if (city_data.country) {
+          this.country = city_data.country.names.en;
```
I see `names.en` which I assume is "english" - what other names are given? It'd be very nice to localize city and country names in our outgoing emails.

shane-tomlinson commented on tasks/build.js:
```diff
@@ -0,0 +1,12 @@
+
+  grunt.registerTask('build',
+    'Build client',
+    ['lint', 'requirejs']);
```
I don't see a requirejs task anywhere. If not needed, might as well remove this file.

shane-tomlinson commented on tasks/eslint.js:
```diff
@@ -0,0 +1,19 @@
+
+module.exports = function (grunt) {
+  'use strict';
+  grunt.config('eslint', {
```
Space above this line please.

shane-tomlinson commented on test/fxa-geodb.js:
```diff
@@ -0,0 +1,94 @@
+  var ip;
+
+  it('returns a promise when called', function () {
+    assert.isTrue(typeof geoDb('12.23.34.45').then === 'function', 'Promise not returned');
```
Chai has an `assert.isFunction` I believe.

`Promise not returned` or `promise is returned`?

shane-tomlinson commented on test/fxa-geodb.js:
```diff
@@ -0,0 +1,94 @@
+        assert.deepEqual(location.ll, ll, 'LatLong not returned correctly');
+        assert.equal(location.time_zone, 'America/Los_Angeles', 'Timezone not returned correctly');
+      }, function (err) {
+        assert.equal(err.message, ERRORS.UNABLE_TO_FETCH_DATA, 'Incorrect error message');
```
IIUC, you should never expect the fail callback to be invoked, it should be an `assert.fail`.

shane-tomlinson commented on test/fxa-geodb.js:
```diff
@@ -0,0 +1,94 @@
+        assert.equal(location.continent, 'North America', 'Continent not returned correctly');
+        assert.equal(typeof location.time_zone, 'undefined', 'Timezone not undefined');
+      }, function (err) {
+        assert.equal(err.message, ERRORS.UNABLE_TO_FETCH_DATA, 'Incorrect error message');
```
Same comment here about the fail callback should not be invoked, this should be an assert.fail.

shane-tomlinson commented on README.md:
````diff
@@ -1,2 +1,49 @@
+On successful resolution of the promise, the `location` object has the following data:
+
+```JavaScript
+accuracy: 'accuracy-radius-in-km', // 5
````
Is this a number or a string?

shane-tomlinson commented on README.md:
```diff
@@ -1,2 +1,49 @@
+continent: 'human-readable-continent-name', // North America
+country: 'human-readable-country-name', // USA
+ll: {
+    latitude: 'latitude-in-decimal', // 37.386
```
Are these numbers or strings?

shane-tomlinson commented on README.md:
```diff
@@ -1,2 +1,49 @@
+--
+
+### Testing
+Mocha Tests are provided inside `test/fxa-geodb.js`. To run the tests, simply call `npm test`.
```
`provided inside` => `located in the test subdirectory`. If you add more tests for the cron function, you'll have two test files, so might as well just put `test`.

Can you remove `simply` from `simply call` here and below?

shane-tomlinson commented on package.json:
```diff
@@ -0,0 +1,45 @@
+  },
+  "scripts": {
+    "setup": "npm install && bower install",
+    "build": "grunt build && node scripts/max-loader.js &",
```
`build` seems like the wrong command because we aren't actually building anything.

TDA commented on scripts/maxmind-db-downloader.js:
```diff
@@ -0,0 +1,126 @@
+  // By default, we do not setup autoUpdate, needs to be
+  // done through options when library is imported, or manually
+  // by running npm run-scripts update
+  // maxmindDbDownloader.setupAutoUpdate('30 30 1 * * 3', remainingDownloads);
```
Not sure how we want the `setupAutoUpdate` called, need clarification on whether the user is allowed to configure this when importing the `fxa-geodb` using require (this might not be a good idea, as the cron job will stop executing when the script that called it stops execution), or whether user will run a special script e.g. `npm run-scripts auto-update` to set up the cron job, where auto-update will have the above code.
If user is going to run this from command-line, then what options are allowed to be customized/configured?

@shane-tomlinson 
@vladikoff

shane-tomlinson replied:
I'm going to admit, auto-updating scares the hell out of me. It introduces a vector by which the auth-server/mailer could go down if a corrupt DB is downloaded. For production, can we disable this feature and only download when the new stack is put together?

vladikoff commented on test/maxmind-db-downloader.js:
```diff
@@ -0,0 +1,85 @@
+
+  describe('setupAutoUpdate', function () {
+    it('auto update calls startDownload correctly', function (done) {
+      // test takes slightly over 60 seconds, set
```
We cannot have tests take 1 minute, can we just adjust the cron to be <10 seconds?

TDA replied:
@vladikoff Sure, I can set it to run every second, and after x seconds check whether it was called x times. Sounds good?

vladikoff commented on scripts/maxmind-db-downloader.js:
```diff
@@ -0,0 +1,126 @@
+    cronTiming = cronTiming || DEFAULTS.DEFAULT_CRON_TIMING;
+    timeZone = timeZone || DEFAULTS.DEFAULT_TIMEZONE;
+    var self = this;
+    // by default run periodically on Wednesday at 01:30:30.
```
seems like this comment is not relevant anymore?

TDA replied:
@vladikoff We still default to it, the DEFAULT_CRON_TIMING is set to wednesdays at 1.30am

TDA commented on package.json:
```diff
@@ -0,0 +1,50 @@
+  "scripts": {
+    "cover": "istanbul cover _mocha",
+    "lint": "grunt lint",
+    "postinstall": "node scripts/max-loader.js",
```
Need to change this to point to the right file @TDA

pdehaan commented on lib/defaults.js:
```diff
@@ -0,0 +1,18 @@
+
+DEFAULTS.DEFAULT_TARGET_DIR_PATH = path.join(__dirname, '..', DEFAULTS.DEFAULT_TARGET_DIR_NAME);
+DEFAULTS.DEFAULT_DB_PATH = path.join(DEFAULTS.DEFAULT_TARGET_DIR_PATH, DEFAULTS.DEFAULT_TARGET_FILE_NAME);
+DEFAULTS.DEFAULT_SOURCE_FILE_PATH = path.join(__dirname, '..', DEFAULTS.DEFAULT_SOURCE_FILE_NAME);
```
FWIW, these feel very verbose... Why not just `DEFAULTS.SOURCE_FILE_NAME`?

pdehaan replied:
Or not sure if it makes any sense to put all these in some sort of **convict** config.

jrgm replied:
Total bikeshedding from a hotel room in Canada, eh?, but write these as if they are `consts` (even if spelled `var` for now). JS hash members are mutable, and these values are not intended to be mutable.

jrgm replied:
Or, I guess what I mean is ALL_CAPS is traditionally used by convention for "const" values. Using them in struct members, is mixing metaphors (or is it similies in this case).

jrgm replied:
Oh, my bad. This is the `module.exports` value. Nevermind. It's the air hear in the north, I swear.

pdehaan commented on package.json:
```diff
@@ -0,0 +1,50 @@
+  "name": "fxa-geodb",
+  "version": "0.0.2",
+  "private": true,
+  "description": "Firefox AccountsGeoDB Repo for Geolocation based services",
```
Probably want a space between "Accounts" and "GeoDB".

pdehaan commented on package.json:
```diff
@@ -0,0 +1,50 @@
+  "devDependencies": {
+    "chai": "3.5.0",
+    "eslint-config-fxa": "2.1.0",
+    "grunt": "0.4.5",
```
Update to newer Grunt, or keeping this in sync with something else FxA related?

pdehaan commented on scripts/maxmind-db-downloader.js:
```diff
@@ -0,0 +1,126 @@
+      stream.pipe(zlib.createGunzip()).pipe(fs.createWriteStream(targetFilePath)).on('finish', function (err) {
+        // forces overwrite, even if file exists already
+        if (err) {
+          log.error(err);
```
Do you need `callback(err)` here?

TDA replied:
Only logging errors I think @pdehaan

vladikoff replied:
does this mean if there is an error then download will never callback?

pdehaan commented on src/fxa-geodb.js:
```diff
@@ -0,0 +1,87 @@
+var path = require('path');
+
+var DEFAULTS = require(path.join('..', 'lib', 'defaults'));
+var ERRORS = require(path.join('..', 'lib', 'errors'));
```
Why not just `require('../lib/errors')`? I think Node normalizes those automatically for different OSes anyways.

TDA replied:
cuz @shane-tomlinson

jrgm replied:
I quoted the relevant nodejs doc about special handling of `require` ;-)

TDA replied:
@jrgm My bad, I think i lost the comment in the overwrite pushes 👍  Thank you! changing back!

pdehaan commented on package.json:
```diff
@@ -0,0 +1,50 @@
+{
+  "name": "fxa-geodb",
+  "version": "0.0.2",
+  "private": true,
```
Probably remove `"private": true` if we need to publish this to npm.

pdehaan commented on package.json:
```diff
@@ -0,0 +1,50 @@
+    "cover": "istanbul cover _mocha",
+    "lint": "grunt lint",
+    "postinstall": "node scripts/max-loader.js",
+    "setup": "npm install && bower install",
```
We using bower? This is only reference I see.

TDA replied:
Good catch @pdehaan We dont, this was leftover from me incorrecly assuming this was on client side.

vladikoff replied:
Yeah we can just remove the whole `setup` step

pdehaan commented on src/fxa-geodb.js:
```diff
@@ -0,0 +1,87 @@
+
+  if (locationData.country) {
+    this.country = this.getLocaleSpecificLocationString(locationData.country, userLocale);
+  }
```
I think you can get `state` information too, via `subdivisions` array.

shane-tomlinson replied:
This seems reasonable to add, because chances are we are going to need it.

pdehaan commented:
This Geo stuff is pretty fun. @TDA, here's a really crude example of me grabbing the state/province information using `subdivisions[]` array (note that it probably lacks a ton of extra data validation):

```js
var path = require('path');
var maxmind = require('maxmind');

var db = path.join(__dirname, 'data', 'GeoLite2-City.mmdb');

var lookup = maxmind.open(db);

var data = getCityData('76.22.116.72', 'en');

console.log(JSON.stringify(data, null, 2));


function getCityData(ip, locale) {
  locale = locale || 'en';

  var getLocaleSpecificLocationString = function (obj, locale) {
    if (!obj || !obj.names) {
      return null;
    }
    if (obj.names.hasOwnProperty(locale)) {
      return obj.names[locale];
    }
    return obj.names.en;
  };

  var data = lookup.get(ip);
  if (!data) {
    throw new Error("Unable to get geo data for " + ip);
  }

  delete data.location.metro_code;

  return {
    city: getLocaleSpecificLocationString(data.city, locale),
    continent: getLocaleSpecificLocationString(data.continent, locale),
    continent_code: data.continent && data.continent.code,
    country: getLocaleSpecificLocationString(data.country, locale),
    country_code: data.country && data.country.iso_code,
    location: data.location,
    postal: data.postal && data.postal.code,
    state: getLocaleSpecificLocationString(data.subdivisions && data.subdivisions[0], locale),
    state_code: data.subdivisions && data.subdivisions[0].iso_code
  };
}
```

Using my IP address, here's what I get:

```json
{
  "city": "Sunnyvale",
  "continent": "North America",
  "continent_code": "NA",
  "country": "United States",
  "country_code": "US",
  "location": {
    "accuracy_radius": 10,
    "latitude": 37.3697,
    "longitude": -122.0214,
    "time_zone": "America/Los_Angeles"
  },
  "postal": "94086",
  "state": "California",
  "state_code": "CA"
}
```

vladikoff commented on README.md:
```diff
@@ -1,2 +1,66 @@
+Provides a wrapper around [node-maxmind] (https://github.com/runk/node-maxmind) for getting the location data in an easy-to-read format.
+
+### Getting started
+Clone/fork the repo and run `npm i`. Then, include the module in your source file, like so:
```
these instructions should be npm instructions rather than `cloning and forking`

vladikoff commented on examples/fxa-geodb-example.js:
```diff
@@ -0,0 +1,22 @@
+var path = require('path');
+
+(function() {
+  'use strict';
```
the example can be simplified:

1. remove `use strict`
2. remove function IIFE
3. the require statement should just point to https://github.com/TDA/fxa-geodb/blob/master/package.json#L6 without `path.join` using a relative path.

vladikoff replied:
add an inline note re: how to include the module via npm before the require statement on line 10.

vladikoff commented on examples/fxa-geodb-example.js:
```diff
@@ -0,0 +1,22 @@
+  });
+  // New York timezone IP: 128.192.8.8
+  // Beijing: 123.121.221.194
+  // Undefined tz : 64.11.221.194
```
nit: extra space after `tz`

vladikoff commented on lib/errors.js:
```diff
@@ -0,0 +1,10 @@
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+var ERRORS = {
+  'IS_INVALID': '"IP" is invalid',
```
nit: not sure if "IP" needs to be in quotes

vladikoff commented on scripts/maxmind-db-downloader.js:
```diff
@@ -0,0 +1,126 @@
+      if (err) {
+        log.error(err);
+      } else {
+        if (process.env.NODE_ENV !== 'test') {
```
what is the purpose of checking !== test ?

TDA replied:
@vladikoff To make sure the tests arent cluttered with the log messages, as mozlog prints to console otherwise.

shane-tomlinson replied:
This is similar to my comment above, just extract all these checks/log.info's into a common function and reduce the duplication.

vladikoff commented on src/fxa-geodb.js:
```diff
@@ -0,0 +1,87 @@
+    return new Promise(function (resolve, reject) {
+      // check if ip is valid
+      if (! maxmind.validate(ip)) {
+        reject({
```
nit: just `return reject`.... might be cleaner

vladikoff commented on src/fxa-geodb.js:
```diff
@@ -0,0 +1,87 @@
+  options = options || {};
+  var dbPath = options.dbPath || DEFAULTS.DEFAULT_DB_PATH;
+
+  var dbLookup, locationData;
```
do you need to define `locationData` here or can it be done below when you fetch it

shane-tomlinson replied:
Place these declarations on separate lines please:

```js
var dbLookup;
var locationData;
```

vladikoff commented on src/fxa-geodb.js:
```diff
@@ -0,0 +1,87 @@
+        return;
+      }
+
+      locationData = dbLookup.get(ip);
```
ref: https://github.com/mozilla/fxa-geodb/pull/1/files#r70446162

vladikoff commented on src/fxa-geodb.js:
```diff
@@ -0,0 +1,87 @@
+        reject({
+          message: ERRORS.UNABLE_TO_FETCH_DATA
+        });
+        return;
```
`return reject`

vladikoff commented on src/fxa-geodb.js:
```diff
@@ -0,0 +1,87 @@
+  } catch (err) {
+    // if it failed with primary database
+    // then quit with error
+    throw ERRORS.UNABLE_TO_FETCH_DATA;
```
should this Promise reject with error instead of throwing unlike other errors?

TDA replied:
From mtg with Shane from last week: If theres an error thrown it will be seen by jbuck or jrgm.

shane-tomlinson replied:
This seems like a reasonable path to me. `UNABLE_TO_FETCH_DATA` doesn't seem like the right error though, FETCH indicates a problem fetching. This is a problem opening.

vladikoff commented on src/fxa-geodb.js:
```diff
@@ -0,0 +1,87 @@
+  this.getLocaleSpecificLocationString = function (locationObject, userLocale) {
+    // if we have the user's locale specific name, return that,
+    // else return 'en' - english. The user's locale refers to
+    // the locale he signed up from, not the locale of the ip
```
`he` -> `they`

vladikoff commented on src/fxa-geodb.js:
```diff
@@ -0,0 +1,87 @@
+  };
+};
+
+function Location(locationData, userLocale) {
```
is none of the `userLocale` stuff documented? Should the args to `Location` be `ip, options` because `userLocale` seems really secondary.

TDA replied:
> Should the args to `Location` be `ip, options`

Did you mean `locationData, options` ? If so, I can change that.
@vladikoff

TDA commented:
@vladikoff Done with the changes you and @pdehaan suggested.

shane-tomlinson commented on tasks/eslint.js:
```diff
@@ -0,0 +1,20 @@
+  grunt.config('eslint', {
+    app: {
+      eslintrc: '.eslintrc',
+      src: ['Gruntfile.js', 'tasks/*.js', 'src/*.js', 'scripts/*.js', 'lib/*.js']
```
Seems like we could do something like:

```js
src: [
  '**/*.js',
  '!node_modules/'
]
```

My grunt-foo is not as good as other people on this team though.

TDA replied:
With the exclusions, this would still have 4 elements: 

```
src: [
  '**/*.js',
  '!node_modules/**/*',
  '!examples/**/*,
  '!test/**/*'
]
```

Since we clubbed src and lib into 1 folder anyway, I think having 4 folders specified explicitly is better.

pdehaan replied:
@TDA, 

1. ESLint should be smart enough to not lint ./node_modules/ by default.
1. Re: `!examples/**/*`, this is probably what .eslintignore is for. Just add "examples" in that file and it will recursively ignore everything. We could also just inline disable the `no-console` rule, if that is the only violator.
1. Re: `!test/**/*`, I guess you maybe have to leave that as an exclude since you explicitly have it as a target below on line 13.

I'd think that since everything uses the same global .eslintrc, probably just easiest to have one target and merge everything together.

Just my $0.01 CAD.

shane-tomlinson commented on tasks/lint.js:
```diff
@@ -0,0 +1,12 @@
+  grunt.registerTask('lint',
+    'Alias for eslint',
+    ['eslint']);
+
```
Pedant alert: ditch the extra line.

shane-tomlinson commented on scripts/maxmind-db-downloader.js:
```diff
@@ -0,0 +1,126 @@
+
+var async = require('async');
+var CronJob = require('cron').CronJob;
+var DEFAULTS = require(path.join('..', 'lib', 'defaults'));
```
As @jrgm pointed out elsewhere, `require` is a special flower and doesn't need `path.join`. My mistake.

shane-tomlinson commented on scripts/maxmind-db-downloader.js:
```diff
@@ -0,0 +1,126 @@
+    // push each file-load-function onto the remainingDownloads queue
+    for (var source in sources) {
+      var url = sources[source];
+      if (process.env.NODE_ENV !== 'test') {
```
I see the pattern of checking NODE_ENV for 'test', then logging, repeated all over the place. I would extract that duplicate functionality into a function called `log` that does the check for you.

shane-tomlinson commented on scripts/maxmind-db-downloader.js:
```diff
@@ -0,0 +1,126 @@
+    targetDirName = targetDirName || DEFAULTS.DEFAULT_TARGET_DIR_NAME;
+    var targetDirPath = path.join(__dirname, '..', targetDirName);
+    // create db folder
+    mkdirp.sync(targetDirPath);
```
What happens if the target directory already exists, does this succeed or fail?

shane-tomlinson commented on scripts/maxmind-db-downloader.js:
```diff
@@ -0,0 +1,126 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+var path = require('path');
```
Put this in strict mode, I think you'll find there are several variables that miss a `var`

shane-tomlinson commented on scripts/maxmind-db-downloader.js:
```diff
@@ -0,0 +1,126 @@
+
+  this.download = function (url, targetFilePath) {
+    // closure to separate multiple file-download
+    return function downloadFunctor(callback) {
```
Is this a Functor or a simple function?

shane-tomlinson replied:
Why not return a promise instead of a function that takes a callback?

shane-tomlinson commented on src/fxa-geodb.js:
```diff
@@ -0,0 +1,87 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
```
I realize I'm being super pedantic here, most of our group's standalone node libraries use the convention that the code is in the `lib` subdir. Consistency makes it easier to understand the source tree.

TDA replied:
@shane-tomlinson I was simply copying the structure we have for fxa-easter-egg and fxa-password-strength-checker, should this be changed to `lib`? If so, then where do I put the `errors.js` and `defaults.js` files?

shane-tomlinson commented on scripts/maxmind-db-downloader.js:
```diff
@@ -0,0 +1,126 @@
+  };
+};
+
+if (require.main === module) {
```
I personally would move most of the downloader code into a module in `src` [or `lib`](https://github.com/mozilla/fxa-geodb/pull/1/files/75c02a076917d3db24e729d3c31dbf5652208739#r70795185), which would export MaxmindDbDownloder. The only thing that'd remain in this module are lines 111-124. The separation of concerns would make each easier to reason about.

shane-tomlinson commented on test/maxmind-db-downloader.js:
```diff
@@ -0,0 +1,87 @@
+      targetDirPath = maxmindDbDownloader.createTargetDir('test-db');
+      assert.equal(targetDirPath, path.join(__dirname, '..', 'test-db'), 'Directory path is correct');
+      assert.isTrue(fs.statSync(targetDirPath).isDirectory(), 'Directory was created');
+    });
```
Add a test if the directory already exists.

shane-tomlinson commented on test/maxmind-db-downloader.js:
```diff
@@ -0,0 +1,87 @@
+      remainingDownloads = maxmindDbDownloader.setupDownloadList(path.join('..', 'sources.json'), targetDirPath);
+      maxmindDbDownloader.startDownload(remainingDownloads);
+      assert.isTrue(async.parallel.called, 'Async was called');
+      assert.isTrue(async.parallel.calledWith(remainingDownloads), 'Async was called with the array');
```
What happens if the download list is empty?

TDA replied:
FWIW test still passes, as it does get called with an empty array.

shane-tomlinson commented on scripts/maxmind-db-downloader.js:
```diff
@@ -0,0 +1,126 @@
+    return function downloadFunctor(callback) {
+      var stream = request(url);
+      stream.pipe(zlib.createGunzip()).pipe(fs.createWriteStream(targetFilePath)).on('finish', function (err) {
+        // forces overwrite, even if file exists already
```
Does this comment apply to the line below or above? If above, move the comment to line 66.

shane-tomlinson commented on scripts/maxmind-db-downloader.js:
```diff
@@ -0,0 +1,126 @@
+  };
+
+  this.startDownload = function (remainingDownloads) {
+    async.parallel(remainingDownloads, function (err) {
```
If [this.download returned a promise](https://github.com/mozilla/fxa-geodb/pull/1/files/75c02a076917d3db24e729d3c31dbf5652208739#r70796126), you could ditch the `async` library and just use [Promise.all](http://bluebirdjs.com/docs/api/promise.all.html). It would also be useful for `startDownload` to return a promise that resolves when all downloads are complete.

`startDownload` seems like a bad name for this function, it seems like it should be `downloadAll` or something similar. And `remainingDownloads` could also be clarified by renaming to something like `downloadPromises`

shane-tomlinson commented on scripts/maxmind-db-downloader.js:
```diff
@@ -0,0 +1,126 @@
+    return remainingDownloads;
+  };
+
+  this.download = function (url, targetFilePath) {
```
`download` implies the download is started, whereas a function is returned that must be invoked. Maybe rename this to `createDownloadFunction` or possibly return a promise, or return a function that returns a promise.

pdehaan commented:
Do we have tests to check ipv4 vs ipv6 as well as results from a lot of different locations so we know if we're getting back the expected data from Singapore and places like that?
It'd be too bad if we were sending out emails with a bunch of "undefined" in the location data.

shane-tomlinson commented:
> Do we have tests to check ipv4 vs ipv6

Do we ever get ipv6 addresses?

rfk commented:
I agree with @shane-tomlinson that auto-update is scary and we need to be very careful that e.g. maxmind's website going down doesn't leave our servers in a bad state.

I think it'd be useful to write out a description of the update process at a high level in the README, including what happens if various errors occur.  One example: it's not clear to me how the automatic update actually interacts with running code, if at all.  AFAICT if I've done this in my code:

```
var geoDb = require('fxa-geodb')({
        dbPath: 'cities-db.mmdb'
});
```

Then I have a `geoDb` object that that has slurped a copy of `cities-db.mmdb` into memory.  If I've got automatic updates going on, how do I ensure that my code picks up the new version of the db?

rfk commented on scripts/maxmind-db-downloader.js:
```diff
@@ -0,0 +1,128 @@
+    // closure to separate multiple file-download
+    return function downloadFunctor(callback) {
+      var stream = request(url);
+      stream.pipe(zlib.createGunzip()).pipe(fs.createWriteStream(targetFilePath)).on('finish', function (err) {
```
Does this overwrite the file in-place?  If so, that seems very fragile in the face of e.g. network outages.  Could we stream it into a temporary file, then replace the original with an atomic `fs.rename(tempFile, targetFilePath)` once we know the download was successful?

TDA replied:
@rfk I did, in fact, have a mechanism to move the existing file to a backup and then try downloading a new one. Not sure why that was asked to be removed.

rfk replied:
>  move the existing file to a backup

FWIW I don't think this is the right approach, because it can leave you without the original file.  I would instead download the *new* file to some temporary location like `cities.mmdb.new` and then once we're sure it's downloaded correctly, only then overwrite the existing file.

But I don't have all the context here from previous discussions, I could easily be missing something.

pdehaan replied:
Just because design by committee is fun ... 
We could also use https://dev.maxmind.com/geoip/geoipupdate/ or even publish the .mmdb as a [private/scoped npm package](https://docs.npmjs.com/private-modules/intro) (ie: @mozilla/maxmind-geolite2-city). We can then manually download+verify checksums and publish the 60MB module to npm as much as we want.

Then it may theoretically be as "simple" as:

```js
var geoDb = require('fxa-geodb')({
  dbPath: require.resolve('@mozilla/maxmind-geolite2-city')
});
```

TDA replied:
So:
On the downloader end:
copy cities.db --> cities-backup.db
download new cities.db

On the src side where we consume,
try opening cities.db, if failure --> try with cities-backup.db

This was what was in place already, however, I do see advantages in copying to a temp location followed by a check for whether the file downloaded correctly. IIRC, Shane told me that comparing checksum and the like was adding too much complexity, so I will defer this to him.

rfk replied:
I agree that's too much complexity on the src side where you consume, so my proposal is:

On the downloader end:
* download cities-new.db
* validate that you can open it correctly, if that's a thing that it's possible to do cheaply
* atomically rename cities-new.db over the top of cities.db

On the src side where we consume:
* do exactly what you're doing now - just open cities.db and use it, erroring out if it doesn't exist

It isolates all the complexity of making sure we don't screw up the state to downloader, which can be run or not under ops control.

pdehaan commented:
> Do we ever get ipv6 addresses?

\*shrug\* I think @jrgm was just telling me we don't get any currently due to AWS.
_But_ I did just verify that the maxmind **GeoLite2-City.mmdb** does work for both IPv4 and IPv6 address (in case we ever do encounter them in the future).

TDA commented:
@pdehaan maxmind.validate() takes in both ipv4 and ipv6, IIUC, that functionality is already tested downstream in the node-maxmind repos?

TDA commented:
Changes made:
1. Added `state` data
2. Changed callbacks to promises
3. Added tests for existing directory
4. Changed tests for promises
5. Added a loghelper to separate that logic from code
6. Split the downloader code into 2 files, and moved all source files to /lib

Whats left:
1. Update mechanism
2. README update

@vladikoff 
@shane-tomlinson

pdehaan commented on examples/fxa-geodb-example.js:
```diff
@@ -0,0 +1,19 @@
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+var path = require('path');
```
PROTIP: You can add `/* eslint-disable strict, no-console */` after the license there and it will disable those 2 rules for this entire file. Then you can lint the examples/* folder too.

Conversely, you can add a nested examples/.eslintrc file which just turns those 2 rules off (and will still inherit the parent rules):

```yaml
rules:
  no-console: 0
  strict: 0
```

pdehaan commented on lib/defaults.js:
```diff
@@ -0,0 +1,18 @@
+
+DEFAULTS.DEFAULT_TARGET_DIR_PATH = path.join(__dirname, '..', DEFAULTS.TARGET_DIR_NAME);
+DEFAULTS.DEFAULT_DB_PATH = path.join(DEFAULTS.DEFAULT_TARGET_DIR_PATH, DEFAULTS.TARGET_FILE_NAME);
+DEFAULTS.DEFAULT_SOURCE_FILE_PATH = path.join(__dirname, '..', DEFAULTS.SOURCE_FILE_NAME);
```
FWIW, you may need/want to tweak the `DEFAULTS_` prefix in these last 3 items:

```json
{
  "CRON_TIMING": "30 30 1 * * 3",
  "SOURCE_FILE_NAME": "sources.json",
  "TARGET_DIR_NAME": "db",
  "TARGET_FILE_NAME": "cities-db.mmdb",
  "TIMEZONE": "America/Los_Angeles",
  "DEFAULT_TARGET_DIR_PATH": "/Users/pdehaan/dev/tmp/del/fxa-geodb/db",
  "DEFAULT_DB_PATH": "/Users/pdehaan/dev/tmp/del/fxa-geodb/db/cities-db.mmdb",
  "DEFAULT_SOURCE_FILE_PATH": "/Users/pdehaan/dev/tmp/del/fxa-geodb/sources.json"
}
```

The variables are still [a semi-verbose] `DEFAULTS.DEFAULT_SOURCE_FILE_PATH`;

jrgm commented on lib/fxa-geodb.js:
```diff
@@ -0,0 +1,88 @@
+  } catch (err) {
+    // if it failed with primary database
+    // then quit with error
+    throw ERRORS.UNABLE_TO_OPEN_FILE;
```
Just to add to @vladikoff's comment in https://github.com/mozilla/fxa-auth-server/pull/1334/files#r71040754, failure to get geolocation information should not be treated as a fatal error. The notification **must** be sent, and deal with missing location data in a way that makes the most sense to the user (which is what will have to happen anyways, since a given IP address is not guaranteed to have location information, or may not even even be accurate).

However, we should work on a way to scream loudly that we are not getting geolocation "hits", whether the cause is a missing or corrupt or invalid data file, or an error in this module, or un-resolvable IP addresses.

shane-tomlinson replied:
Thanks for the input @jrgm, I was the one that said the fail entirely here.

TDA replied:
From mtg with Shane: fail silently until the lookup is called, and reject with a promise there if it was null.

TDA commented:
Changes made:
1. Update mechanism as suggested by @rfk (Create a temp file first, verify that file can be searched for locations, then do atomic rename)
2. Remove `throw error` , and instead use rejected promise.

Whats left:
1. README update (update mechanism, usage, data returned, and examples)
@vladikoff

vladikoff commented on README.md:
````diff
@@ -1,2 +1,75 @@
+
+```JavaScript
+var geoDb = require('fxa-geodb')({
+	dbPath: 'cities-db.mmdb' // Defaults to fxa-geodb/db/cities-db.mmdb
````
needs indent 2 spaces

vladikoff commented on README.md:
```diff
@@ -1,2 +1,75 @@
+      }, function (err) {
+        // rejected :(
+        // Uh-oh error
+      });
```
needs proper indent above

vladikoff commented on lib/fxa-geodb.js:
```diff
@@ -0,0 +1,92 @@
+module.exports = function (options) {
+  'use strict';
+
+  options = options || {};
```
@TDA these options are not documented anywhere, I suggest JSDOC them and generate an MD file, so we don't have to manually update docs all the time

Example: https://github.com/vladikoff/foxfire/blob/master/docs.md and https://github.com/vladikoff/foxfire/blob/master/Gruntfile.js#L4

vladikoff replied:
This can be a follow up issue

vladikoff commented on lib/fxa-geodb.js:
```diff
@@ -0,0 +1,92 @@
+    dbLookup = maxmind.open(dbPath);
+  } catch (err) {
+    // if it failed with primary database
+    // then do nothing
```
> then do nothing

A better comment would probably be `then reject the promise below`

vladikoff replied:
@TDA  How would the deploy team know that something went wrong with the `dbPath`?

TDA replied:
@vladikoff The error logged will be `'UNABLE_TO_OPEN_FILE'`, as that is the error message returned by the check right below.

vladikoff commented on lib/fxa-geodb.js:
```diff
@@ -0,0 +1,92 @@
+
+  return function (ip, options) {
+    options = options || {};
+    var userLocale = options.userLocale || 'en';
```
`'en'` can be part of DEFAULTS

vladikoff commented on lib/fxa-geodb.js:
```diff
@@ -0,0 +1,92 @@
+  };
+};
+
+function Location(locationData, userLocale) {
```
can Location be loaded from its own file via `var Location = require('./location');` ?

vladikoff commented on lib/fxa-geodb.js:
```diff
@@ -0,0 +1,92 @@
+      // latitude, and longitude, and timezone
+      var location = new Location(locationData, userLocale);
+
+      resolve(location);
```
add `return` here to be consistent with `rejects`

vladikoff commented on lib/maxmind-db-downloader.js:
```diff
@@ -0,0 +1,129 @@
+    var targetDirPath = path.join(__dirname, '..', targetDirName);
+    // create db folder
+    var createdTargetDirPath = mkdirp.sync(targetDirPath);
+    logHelper('info', 'Download folder is ' + createdTargetDirPath);
```
`folder` -> `directory`

vladikoff commented on lib/maxmind-db-downloader.js:
```diff
@@ -0,0 +1,129 @@
+              dbPath: targetFilePathTemp
+            });
+            // check if lookup works with the downloaded file
+            geoDb('8.8.8.8')
```
need better logs here, maybe `before` and `after` attempt to "test" the new download

vladikoff replied:
also add `8.8.8.8` to DEFAULTS or above as constant

rfk commented on lib/maxmind-db-downloader.js:
```diff
@@ -0,0 +1,129 @@
+                resolve();
+              }, function (err) {
+                // download resulted in an error, do not rename
+                reject(err);
```
nit: delete the known-to-be-invalid temp file rather than leaving it on disk

TDA commented:
Changes made:
1. README update (update mechanism, usage, data returned, and examples)
2. Fixed Indents, added defaults, added logging before and after testing the downloaded file, remove downloaded file on failure.

@vladikoff
@rfk

vladikoff commented:
Ok thanks @TDA ! I merged this and opened a few follow up issues. If anyone else had a comment that was not addressed or required a follow up issue please open one 🍰
